File: rubyrrgarten/TestP1.rb
require_relative 'Shield'
require_relative 'Weapon'
require_relative 'GameState'
require_relative 'Dice'
require_relative 'enums'


puts ".............Probando Weapon y Shield............."

# Creación de instancias de Weapon y Shield
weapon = Irrgarten::Weapon.new(2.0, 5)
shield = Irrgarten::Shield.new(3.0, 4)

# Usar métodos y probar to_s
puts weapon
puts "Weapon attack: #{weapon.attack}"
puts weapon

puts shield
puts "Shield protect: #{shield.protect}"
puts shield

puts "    ..prueba de discard().."

# Probar discard
probabilidad = []  # Array para almacenar probabilidades
for i in 0..Irrgarten::Dice.MAX_USES
  descarta = 0
  veces = 10000
  for j in 0..(veces-1)
    arma = Irrgarten::Weapon.new(2,i)
    if arma.discard
      descarta += 1
    end
  end
  probabilidad << Float(descarta)/Float(veces)
end

(0..5).each do |i|                  # MAX_USES = 5
  puts "    p(discard() == true) para uses=#{i} -> #{probabilidad[i] * 100}%"
end

puts "\n\n.............Probando Enumerados............."

# Probar enumerados

dir = Irrgarten::Directions::UP
character = Irrgarten::GameCharacter::PLAYER
orientation = Irrgarten::Orientation::VERTICAL

puts "Direction: #{dir}"
puts "Character: #{character}"
puts "Orientation: #{orientation}"


puts "\n\n.............Probando Dice............."

times = 1_000_000
promedios = Array.new(11,0.0)

# Probar Dice
for i in 1..times
  promedios[0] += Irrgarten::Dice.random_pos(10)/Float(times)
  promedios[1] += Irrgarten::Dice.who_starts(5)/Float(times)
  promedios[2] += Irrgarten::Dice.random_intelligence/Float(times)
  promedios[3] += Irrgarten::Dice.random_strength/Float(times)
  promedios[4] += (Irrgarten::Dice.resurrect_player ? 1 : 0)/Float(times)
  promedios[5] += Irrgarten::Dice.weapons_reward/Float(times)
  promedios[6] += Irrgarten::Dice.shields_reward/Float(times)
  promedios[7] += Irrgarten::Dice.health_reward/Float(times)
  promedios[8] += Irrgarten::Dice.weapon_power/Float(times)
  promedios[9] += Irrgarten::Dice.shield_power/Float(times)
  promedios[10] += Irrgarten::Dice.uses_left/Float(times)
end


puts "Promedios:"
puts "randomPos:            #{promedios[0]} ≈ 5"
puts "whoStarts:            #{promedios[1]} ≈ 2.5"
puts "randomIntelligence:   #{promedios[2]} ≈ 5"
puts "randomStrength:       #{promedios[3]} ≈ 5"
puts "resurrectPlayer:      #{promedios[4]} ≈ 0.3"
puts "weaponsReward:        #{promedios[5]} ≈ 1"
puts "shieldsReward:        #{promedios[6]} ≈ 1.5"
puts "healthReward:         #{promedios[7]} ≈ 2.5"
puts "weaponPower:          #{promedios[8]} ≈ 1.5"
puts "shieldPower:          #{promedios[9]} ≈ 1"
puts "usesLeft:             #{promedios[10]} ≈ 2.5"

File: javargarten/src/TestP1.java
import irrgarten.Dice;
import irrgarten.GameState;
import irrgarten.Shield;
import irrgarten.Weapon;
import irrgarten.enums.Directions;
import irrgarten.enums.Orientation;
import irrgarten.enums.GameCharacter;


public class TestP1 {
    public static void main(String[] args) {

        System.out.println(".............Probando Weapon y Shield.............");
        // Creación de instancias de Weapon y Shield
        Weapon weapon = new Weapon(2.0f, 5);
        Shield shield = new Shield(3.0f, 4);

        // Usar métodos y probar toString
        System.out.println(weapon);
        System.out.println("Weapon attack: " + weapon.attack());
        System.out.println(weapon);

        System.out.println(shield);
        System.out.println("Shield protect: " + shield.protect());
        System.out.println(shield);

        System.out.println("    ..prueba de discard()..");
        // Probar discard
        float[] probabilidad = new float[5+1];  // MAX_USES = 5
        for (int i = 0; i <= 5; i++){           // MAX_USES = 5
            int descarta = 0;
            float veces=10000000;
            for (int j = 0; j < veces; j++){
                Weapon arma = new Weapon(2,i);
                if (arma.discard()) descarta++;
            }
            probabilidad[i] = (float)descarta/veces;
        }
        for (int i = 0; i <= 5; i++){          // MAX_USES = 5
            System.out.println("    p(discard() == true) para uses="+i+" -> " + probabilidad[i]*100 + "%");
        }

        System.out.println();System.out.println();
        System.out.println(".............Probando Enumerados.............");
        // Probar enumerados
        Directions dir = Directions.UP;
        GameCharacter character = GameCharacter.PLAYER;
        Orientation orientation = Orientation.VERTICAL;

        System.out.println("Direction: " + dir);
        System.out.println("Character: " + character);
        System.out.println("Orientation: " + orientation);

        System.out.println();System.out.println();
        System.out.println(".............Probando Dice.............");
        int times = 10000000;
        float [] promedios = new float[11];
        for (int i = 0; i < times; i++) {
            promedios[0] += Dice.randomPos(10)/(float)times;
            promedios[1] += Dice.whoStarts(5)/(float)times;
            promedios[2] += Dice.randomIntelligence()/(float)times;
            promedios[3] += Dice.randomStrength()/(float)times;
            promedios[4] += (Dice.resurrectPlayer() ? 1 : 0)/(float)times;  
            promedios[5] += Dice.weaponsReward()/(float)times;
            promedios[6] += Dice.shieldsReward()/(float)times;
            promedios[7] += Dice.healthReward()/(float)times;
            promedios[8] += Dice.weaponPower()/(float)times;
            promedios[9] += Dice.shieldPower()/(float)times;
            promedios[10] += Dice.usesLeft()/(float)times;
        }
        System.out.println("Promedios:");   
        System.out.println("randomPos:            " + promedios[0] + " ≈ 5");
        System.out.println("whoStarts:            " + promedios[1] + " ≈ 2.5");
        System.out.println("randomIntelligence:   " + promedios[2] + " ≈ 5");
        System.out.println("randomStrength:       " + promedios[3] + " ≈ 5");
        System.out.println("resurrectPlayer:      " + promedios[4] + " ≈ 0.3");
        System.out.println("weaponsReward:        " + promedios[5] + " ≈ 1");
        System.out.println("shieldsReward:        " + promedios[6] + " ≈ 1.5");
        System.out.println("healthReward:         " + promedios[7] + " ≈ 2.5");
        System.out.println("weaponPower:          " + promedios[8] + " ≈ 1.5");
        System.out.println("shieldPower:          " + promedios[9] + " ≈ 1");
        System.out.println("usesLeft:             " + promedios[10] + " ≈ 2.5");

    }


}
